hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/HoodieCreateRecordUtils.scala

```diff
@@ -0,0 +1,285 @@
+            .asInstanceOf[BaseKeyGenerator]
+
+          val dataFileSchema = new Schema.Parser().parse(dataFileSchemaStr)
+          val consistentLogicalTimestampEnabled = parameters.getOrElse(
+            DataSourceWriteOptions.KEYGENERATOR_CONSISTENT_LOGICAL_TIMESTAMP_ENABLED.key(),
+            DataSourceWriteOptions.KEYGENERATOR_CONSISTENT_LOGICAL_TIMESTAMP_ENABLED.defaultValue()).toBoolean
+
+          // handle dropping partition columns
+          it.map { avroRec =>
+            val (hoodieKey: HoodieKey, recordLocation: Option[HoodieRecordLocation]) = HoodieCreateRecordUtils.getHoodieKeyAndMaybeLocationFromAvroRecord(keyGenerator, avroRec,
+              isPrepped)
+            val avroRecWithoutMeta: GenericRecord = if (isPrepped) {
+              HoodieAvroUtils.rewriteRecord(avroRec, HoodieAvroUtils.removeMetadataFields(dataFileSchema))
+            } else {
+              avroRec
+            }
+
+            val processedRecord = if (shouldDropPartitionColumns) {
+              HoodieAvroUtils.rewriteRecord(avroRecWithoutMeta, dataFileSchema)
+            } else {
+              avroRecWithoutMeta
+            }
+
+            val hoodieRecord = if (shouldCombine) {
+              val orderingVal = HoodieAvroUtils.getNestedFieldVal(avroRec, config.getString(PRECOMBINE_FIELD),
+                false, consistentLogicalTimestampEnabled).asInstanceOf[Comparable[_]]
+              DataSourceUtils.createHoodieRecord(processedRecord, orderingVal, hoodieKey,
+                config.getString(PAYLOAD_CLASS_NAME), recordLocation)
+            } else {
+              DataSourceUtils.createHoodieRecord(processedRecord, hoodieKey,
+                config.getString(PAYLOAD_CLASS_NAME), recordLocation)
+            }
+            hoodieRecord
+          }
+        }).toJavaRDD()
+
+      case HoodieRecord.HoodieRecordType.SPARK =>
+        val dataFileSchema = new Schema.Parser().parse(dataFileSchemaStr)
+        val dataFileStructType = HoodieInternalRowUtils.getCachedSchema(dataFileSchema)
+        val writerStructType = HoodieInternalRowUtils.getCachedSchema(writerSchema)
+        val sourceStructType = df.schema
+
+        if (isPrepped) {
+          // For prepped records, check the record schema to make sure it has meta fields set.
+          validateMetaFieldsInSparkRecords(sourceStructType)
+        }
+
+        df.queryExecution.toRdd.mapPartitions { it =>
+          val sparkPartitionId = TaskContext.getPartitionId()
+          val keyGenProps = new TypedProperties(config.getProps)
+          if (autoGenerateRecordKeys) {
+            keyGenProps.setProperty(KeyGenUtils.RECORD_KEY_GEN_PARTITION_ID_CONFIG, String.valueOf(sparkPartitionId))
+            keyGenProps.setProperty(KeyGenUtils.RECORD_KEY_GEN_INSTANT_TIME_CONFIG, instantTime)
+          }
+          val sparkKeyGenerator = HoodieSparkKeyGeneratorFactory.createKeyGenerator(keyGenProps).asInstanceOf[SparkKeyGeneratorInterface]
```

Comment:
lets make this optional as well. prepped flow should have this as empty

Reply:
Done.